lib/Service/Provisioning/Manager.php

```diff
@@ -84,6 +98,71 @@ public function provision(): int {
 		return $cnt;
 	}
 
+	/**
+	 * Delete orphaned aliases for the given account.
+	 *
+	 * A alias is orphaned if not listed in newAliases anymore
+	 * (=> the provisioning configuration does contain it anymore)
+	 *
+	 * Exception for Nextcloud 20: \Doctrine\DBAL\DBALException
+	 * Exception for Nextcloud 21 and newer: \OCP\DB\Exception
+	 *
+	 * @TODO: Change throws to \OCP\DB\Exception once Mail does not support Nextcloud 20.
+	 *
+	 * @throws \Exception
+	 */
+	private function deleteOrphanedAliases(string $userId, int $accountId, array $newAliases): void {
+		$existingAliases = $this->aliasMapper->findAll($accountId, $userId);
+		foreach ($existingAliases as $existingAlias) {
+			if (!in_array($existingAlias->getAlias(), $newAliases, true)) {
+				$this->aliasMapper->delete($existingAlias);
+			}
+		}
+	}
+
+	/**
+	 * Create new aliases for the given account.
+	 *
+	 * Exception for Nextcloud 20: \Doctrine\DBAL\DBALException
+	 * Exception for Nextcloud 21 and newer: \OCP\DB\Exception
+	 *
+	 * @TODO: Change throws to \OCP\DB\Exception once Mail does not support Nextcloud 20.
+	 *
+	 * @throws \Exception
+	 */
+	private function createNewAliases(string $userId, int $accountId, array $newAliases, string $displayName): void {
+		foreach ($newAliases as $newAlias) {
+			try {
+				$this->aliasMapper->findByAlias($newAlias, $userId);
+			} catch (DoesNotExistException $e) {
+				$alias = new Alias();
+				$alias->setAccountId($accountId);
+				$alias->setName($displayName);
+				$alias->setAlias($newAlias);
+				$this->aliasMapper->insert($alias);
+			}
+		}
+	}
+
+	/**
+	 * @throws \Exception if user id was not found in LDAP
+	 */
+	public function ldapAliasesIntegration(Provisioning $provisioning, IUser $user): Provisioning {
+		if ($user->getBackendClassName() !== 'LDAP' || $provisioning->getLdapAliasesProvisioning() === false || empty($provisioning->getLdapAliasesAttribute())) {
+			return $provisioning;
+		}
+
+		if ($this->ldapProviderFactory->isAvailable() === false) {
```

Comment:
Snap! isAvailable is new in 21 :man_shrugging:

Reply:
Do an additional server version comparison here and don't offer the feature for Nextcloud 20 or older. It's a fair tradeoff IMO.

Reply:
Thanks! I added `@psalm-suppress UndefinedInterfaceMethod` to make Psalm on Nextcloud 20 happy. The check for the right Nextcloud version is already done some lines earlier: https://github.com/nextcloud/mail/pull/5198/files#diff-fe646e6e2f7561cf6a8731328355c7432be7647f7114e1619fe2bf5ff2638c42R205. If `getMultiValueUserAttribute` exist `isAvailable` also exist.